h/models/user.py

```diff
@@ -18,27 +19,74 @@
 PASSWORD_MIN_LENGTH = 2
 
 
+def _normalise_username(username):
+    # We normalize usernames by dots and case in order to discourage attempts
+    # at impersonation.
+    return sa.func.lower(sa.func.replace(username,
+                                         sa.text("'.'"),
+                                         sa.text("''")))
+
+
+class UsernameComparator(Comparator):
```

Comment:
That's correct, but only half of the story. Here's the other half. Consider if we had the following code:

```python
@hybrid_property
def username(self):
    return self._username

@username.expression
def username(self):
    return _normalise_username(self._username)
```

Then a query like `session.query(User.username)` would return *normalised* usernames, because `.expression` gets used wherever an expression is needed, including in `SELECT` clauses.

Using a distinct `.expression` and `.comparator` means that you can control comparisons separately, so with the code on this branch you can do something like:

```python
session.query(User.username).filter(User.username.like('david%'))
```

And you'll get back unnormalised usernames that match the filter, where the filter is evaluated against *normalised* usernames.